src/radical/entk/task.py

```diff
@@ -29,1135 +114,227 @@ class Task(object):
-        self._pre_exec   = list()
-        self._executable = ""
-        self._arguments  = list()
-        self._sandbox    = ""
-        self._post_exec  = list()
-        self._stage_on_error = False
-
-        self._lfs_per_process = 0
-        self._mem_per_process = 0
-        self._cpu_reqs        = {'processes'           : 1,
-                                 'process_type'        : None,
-                                 'threads_per_process' : 1,
-                                 'thread_type'         : None}
-        self._gpu_reqs        = {'processes'           : 0,
-                                 'process_type'        : None,
-                                 'threads_per_process' : 0,
-                                 'thread_type'         : None}
-
+        from_dict = from_dict or {}
+        if not isinstance(from_dict, dict):
+            raise ree.TypeError(expected_type=dict,
+                                actual_type=type(from_dict))
 
+        super().__init__()
 
-        # Data staging attributes
-        self._upload_input_data    = list()
-        self._copy_input_data      = list()
-        self._link_input_data      = list()
-        self._move_input_data      = list()
-        self._copy_output_data     = list()
-        self._link_output_data     = list()
-        self._move_output_data     = list()
-        self._download_output_data = list()
+        if not from_dict.get('uid'):
+            self['uid'] = ru.generate_id('task.%(counter)04d', ru.ID_CUSTOM)
+        self['state'] = res.INITIAL
 
-        # Name of file to write stdout and stderr of task
-        self._stdout = ""
-        self._stderr = ""
+        self.update(from_dict)
 
-        # Additional attributes that help in mapping tasks
-        # to cuds and cus to tasks
-        self._path      = None
-        self._exit_code = None
-        self._tags      = None
+    # --------------------------------------------------------------------------
+    #
+    def __eq__(self, other):
 
-        # Keep track of res attained
-        self._state_history = [res.INITIAL]
+        if not isinstance(other, Task):
+            return False
 
-        # Stage and pipeline this task belongs to
-        self._p_stage    = {'uid': None, 'name': None}
-        self._p_pipeline = {'uid': None, 'name': None}
+        return self['uid'] == other['uid']
 
-        # populate task attributes if so requesteed
-        if from_dict:
+    # --------------------------------------------------------------------------
+    #
+    def __ne__(self, other):
 
-            if not isinstance(from_dict, dict):
-                raise ree.TypeError(expected_type=dict,
-                                    actual_type=type(from_dict))
+        return not self == other
 
-            for k,v in from_dict.items():
-                self.__setattr__(k, v)
+    # --------------------------------------------------------------------------
+    #
+    def __hash__(self):
```

Comment:
This doesn't seem right. Instances are mutable, aren't they?

Reply:
Specifically, this implementation allows for the result of `__hash__` to change during the lifetime of the instance (if the 'uid' is set or reset), which is inconsistent with the Python data model, even if it seems like an edge case.

Reply:
From https://docs.python.org/3/reference/datamodel.html#object.__hash__: "If a class defines mutable objects and implements an __eq__() method, it should not implement __hash__(), since the implementation of hashable collections requires that a key’s hash value is immutable (if the object’s hash value changes, it will be in the wrong hash bucket)."

Additionally, the hashability and related discussions regarding `dataclasses` and the various Mappings establish a strong pattern that should be clearly documented if rejected. See also https://docs.python.org/3/library/dataclasses.html?highlight=unsafe_hash

My suggestion is either
* make `uid` immutable if it is the basis of `__hash__`, and clearly document the comparison behaviors, or
* don't implement `__hash__`

I think it would be cleanest to refrain from adding either `__eq__` or `__hash__` with the current semantics.

Reply:
@eirrgang Eric, thank you! That is helpful. We should be able to hash instances of Task since they are collected in `Stage` instance's attribute of type `set`, and thus `uid` will be set during initialization only.